### PR TITLE
Basic authenticator

### PR DESCRIPTION
**Pull request by adam-fowler**
Had to stick this in its own module as it brings together both HummingbirdAuth and Bcrypt

```swift
router.add(
    middleware: BasicAuthenticator { username in
        try await User.query(on: self.fluent.db())
            .filter(\.$name == username)
            .first()
    }
)
```

### CODE REVIEW

File: Sources/HummingbirdBasicAuth/BasicAuthenticator.swift [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    public func authenticate(request: Request, context: Context) async throws -> Repository.User? {
+        // does request have basic authentication info in the "Authorization" header
+        guard let basic = request.headers.basic else { return nil }
+
+        // check if user exists and then verify the entered password against the one stored in the database.
+        // If it is correct then login in user
+        let user = try await users.getUser(named: basic.username)
+        guard let user, let passwordHash = user.passwordHash else { return nil }
+        // Do Bcrypt verify on a separate thread to not block the general task executor
+        guard try await NIOThreadPool.singleton.runIfActive({ Bcrypt.verify(basic.password, hash: passwordHash) }) else { return nil }
```
Comment (Joannis) [inline Sources/HummingbirdBasicAuth/BasicAuthenticator.swift:52]: Do we plan to only support BCrypt for hashing?

Comment (adam-fowler) [reply Sources/HummingbirdBasicAuth/BasicAuthenticator.swift:52]: That's what I was thinking. Maybe use a generic parameter for the encryption method

Comment (adam-fowler) [reply Sources/HummingbirdBasicAuth/BasicAuthenticator.swift]: So I added a PasswordVerifier generic parameter which is defaulted to Bcrypt.

File: Sources/HummingbirdBasicAuth/PasswordVerifier.swift [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import Bcrypt
+import NIOPosix
+
+/// Protocol for password verifier
+public protocol PasswordVerifier: Sendable {
```
Comment (Joannis) [inline Sources/HummingbirdBasicAuth/PasswordVerifier.swift:19]:
```suggestion
public protocol PasswordHashVerifier: Sendable {
```

File: Sources/HummingbirdBasicAuth/PasswordVerifier.swift [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    func verifyPassword(_ text: String, hash: String) async throws -> Bool
```
Comment (Joannis) [inline Sources/HummingbirdBasicAuth/PasswordVerifier.swift:20]:
```suggestion
    func verifyPassword(_ password: String, createsHash hash: String) async throws -> Bool
```

I would love an argument name of some form to clarify to newcomers what the hash's relation is to the passwod

**Review by Joannis — APPROVED** [review]

**Pull request merged**